### DISCUSSION TITLE
How can I define a convection term that includes an independent variable?

### DISCUSSION THREAD

**Discussion by LiuDongDong1989** [usnistgov/fipy#956] [Q&A] [answered]
My equations are as shown in the figure, and while the equations themselves aren't particularly complex, I'm having trouble figuring out how to define them in FiPy. I've tried some code, but I keep getting errors. How can I define the second term in the second equation and the second term in the third equation? 
<img width="1127" alt="截屏2023-09-27 20 29 42" src="https://github.com/usnistgov/fipy/assets/32028558/17acb8ee-d2d6-4498-b7e8-1b439f929484">

The code for the first version is as follows：

    `def CreatEquation(self):
        #h
        self.R = self.P * numerix.cos(self.slopeAngle) - self.i
        self.eq_h = TransientTerm(var=self.h) + ConvectionTerm(var=self.h, coeff=[self.u]) == self.R

        #u
        self.f = 8 * self.g * self.h * numerix.sin(self.slopeAngle) / (self.u**2)
        self.Sfx = (self.u / numerix.sqrt(8 * self.g / self.f))**2 / self.h
        self.eq_u = (TransientTerm(var=self.u) + ConvectionTerm(var=self.u, coeff=[self.u]) + 
                    ConvectionTerm(var=self.h, coeff=[self.g])) == (
                        self.g * (numerix.tan(self.slopeAngle) - self.Sfx) +
                        self.P * numerix.cos(self.slopeAngle) * self.Vm / self.h -
                        self.R * self.u / self.h
                    )
        
        #h & u
        self.eqns = self.eq_h & self.eq_u    `

According the comments from [jeguyer](https://stackoverflow.com/users/2019542/jeguyer), the code for the second version is as follows:

       `self.R = self.P * numerix.cos(self.slopeAngle) - self.i
        self.eqh = TransientTerm(var=self.h) + ConvectionTerm(var=self.h, coeff=self.u * [[1]]) == self.R   
        self.f = 8 * self.g * self.h * numerix.sin(self.slopeAngle) / (self.u**2)
        self.Sfx = (self.u / numerix.sqrt(8 * self.g / self.f))**2 / self.h
        self.equ = ( TransientTerm(var=self.u) + 
                    ConvectionTerm(var=self.u, coeff=(self.u/2) *[[1]])  + 
                    (self.g * self.h.arithmeticFaceValue).divergence ) == (
                    self.g * (numerix.tan(self.slopeAngle) - self.Sfx) +
                    self.P * numerix.cos(self.slopeAngle) * self.Vm / self.h -
                    self.R * self.u /self.h 
                    )
        # h & u
        self.eqns = self.eqh & self.equ`

The above code can run but can not convergence. However, if I use the `ConvectionTerm(coeff=self.g, var=self.h) - ImplicitSourceTerm(coeff=self.g.grad, var=self.h)` to represent the third term of the third equation, the error was:
**_AttributeError: 'float' object has no attribute 'grad'_**. Here, the g was gravitational acceleration(9.8 m^2/s)

[jeguyer](https://stackoverflow.com/users/2019542/jeguyer) told me that **` self.g * self.h.arithmeticFaceValue).divergence` can not represent the third term.**

**Comment by guyer** [comment]
That's not [what I said](https://stackoverflow.com/a/77183365/2019542) your equations should look like. Specifically, you have $g\frac{\partial h}{\partial x} = \frac{\partial}{\partial x}\left(g h\right) - h\frac{\partial g}{\partial x}$, which I said could be written as `ConvectionTerm(coeff=g, var=h) - ImplicitSourceTerm(coeff=g.grad, var=h)`. If $g$ is constant, then you can omit the `ImplicitSourceTerm`. For the `ConvectionTerm`, the `coeff` needs to be a vector. If you have defined `self.g = 9.8`, then you can write `ConvectionTerm(coeff=[self.g], var=h)`. The square brackets around `self.g` mean that this is a Python list with one item in it; FiPy will treat this as a 1D vector.

**Reply by LiuDongDong1989** [reply]
This code`ConvectionTerm(coeff=[self.g], var=h)` is right and can run, but `ConvectionTerm(coeff=self.g * [[1]], var=h)` can not run. What is the differences between them? And `self.g * self.h.arithmeticFaceValue).divergence` can not represent the third term?

**Reply by guyer** [reply, marked as answer]
- If `self.g` is a float, then `self.g * [[1]]` is attempting to multiply a float by a list. Python doesn't know how to do that. The reason that `self.u * [[1]]` works is because `self.u` is a FiPy `CellVariable`. A `CellVariable` "knows" that when it's multiplied with a list, that it should try to interpret that list as a vector. You could get the same behavior with `self.g` if you wrote
    ```python
    self.g = fp.Variable(9.8)
    ConvectionTerm(coeff=self.g * [[1]], var=h)
    ```
- `(self.g * self.h.arithmeticFaceValue).divergence` represents $\nabla\cdot(g h)$ (or $\partial(g h)/\partial x$ in 1D). While, when *and only when* $g$ is constant, this is mathematically equivalent to $g (\partial h/\partial x)$, the expression `(self.g * self.h.arithmeticFaceValue).divergence` is fully explicit in FiPy and will converge poorly compared to `ConvectionTerm(coeff=[self.g], var=h)`.
